### PR TITLE
Refactor module to use IAM policy document instead of JSON templating

### PR DESCRIPTION
**Pull request by dynamike**
This module currently has a bug where if you change any of the AWS logs prefixes (e.g., cloudtrail_logs_prefix) to a root prefix (""), the bucket policy will be broken. In trying to resolve that, I think I've realized we've hit the limits of JSON templating the bucket policy for this module. 

So this PR changes how we generate the bucket policy by using https://www.terraform.io/docs/providers/aws/d/iam_policy_document.html and moves most of the logic to locals. In that refactor, I've added more to tests around root prefixes for each AWS service. I've also tried to document the logic on generating these policies as the module has grown a bit in complexity and there is a mishmash of support for multiple log prefixes vs multiple AWS accounts. 

With that being said, this change shouldn't generate any breaking changes. The upgrade path only require modifying nlb_logs_prefixes and alb_logs_prefixes along with a more correct policy for ALBs and NLBs. Details are in the README.

```
PASS
ok  	github.com/trussworks/terraform-aws-logs/test	2249.807s
```

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -20,255 +16,314 @@ data "aws_caller_identity" "current" {
...
+  alb_include_root_prefix = contains(var.alb_logs_prefixes, "") ? true : false
+
+  # create a list of paths, but remove any prefixes containing "" using compact
+  alb_logs_path_temp = formatlist("%s/AWSLogs", compact(var.alb_logs_prefixes))
+
+  # now append an "AWSLogs" path to the list if alb_include_root_prefix is true
+  alb_logs_path = local.alb_include_root_prefix ? concat(local.alb_logs_path_temp, ["AWSLogs"]) : local.alb_logs_path_temp
+
+  # finally, format the full final resources ARN list
+  alb_resources = sort(formatlist("${local.bucket_arn}/%s/${data.aws_caller_identity.current.account_id}/*", local.alb_logs_path))
```
Comment (eeeady) [inline main.tf]: I'm not sure why you have `${data.aws_caller_identity.current.account_id}` as part of the resource arn for things in the bucket with the prefixes/AWSLogs 

Did you mean to have this there?

File: main.tf [hunk 2]
```diff
@@ -20,255 +16,314 @@ data "aws_caller_identity" "current" {
...
+  # supports logging to multiple prefixes
+  nlb_effect = var.default_allow || var.allow_nlb ? "Allow" : "Deny"
+
+  nlb_include_root_prefix = contains(var.nlb_logs_prefixes, "") ? true : false
+
+  nlb_logs_path_temp = formatlist("%s/AWSLogs", compact(var.nlb_logs_prefixes))
+
+  nlb_logs_path = local.nlb_include_root_prefix ? concat(local.nlb_logs_path_temp, ["AWSLogs"]) : local.nlb_logs_path_temp
+
+  nlb_resources = sort(formatlist("${local.bucket_arn}/%s/${data.aws_caller_identity.current.account_id}/*", local.nlb_logs_path))
```
Comment (eeeady) [inline main.tf]: I feel like I'm missing something here too...

File: main.tf (hunk 1 above)
Comment (dynamike) [reply main.tf]: AWS recommends including the account id in the policy ARN

```
      "Resource": "arn:aws:s3:::bucket_name/prefix/AWSLogs/123456789012/*",
```
https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-access-logs.html

We actually do it for the ELB policy, but not on ALB and NLB.

Comment (eeeady) [reply main.tf]: ooh okay

File: README.md [hunk 3]
```diff
@@ -90,16 +78,16 @@ module "aws_logs" {
...
+      alb_logs_prefixes = [
+       "nlb/alb-hello-world-prod",
+       "nlb/alb-hello-world-staging",
+       "nlb/alb-hello-world-experimental",
+      ]
+      nlb_logs_prefixes = [
+       "alb/nlb-hello-world-prod",
+       "alb/nlb-hello-world-staging",
+       "alb/nlb-hello-world-experimental",
+      ]
```
Comment (cblkwell) [inline README.md]: Uhm...this doesn't seem right. Or at least, this is a very confusing example.

Comment (dynamike) [reply README.md]: whoooops!

File: README.md [hunk 4]
```diff
@@ -153,6 +140,24 @@ module "aws_logs" {
 
 ## Upgrade Paths
 
+### Upgrading from 6.0.0 to 7.x.x
+
+This release simplifies `nlb_logs_prefixes` and `alb_logs_prefixes` to no longer need to pass in generate a formatted list and instead can be referenced as
```
Comment (kilbergr) [inline README.md]:
```suggestion
This release simplifies `nlb_logs_prefixes` and `alb_logs_prefixes` to no longer need to pass in a generated a formatted list and instead can be referenced as
```

**Pull request merged**